src/t_hash.c

```diff
@@ -297,25 +295,406 @@ static void hashDictWithExpireOnRelease(dict *d) {
     ebDestroy(&dictExpireMeta->hfe, &hashFieldExpireBucketsType, NULL);
 }
 
+/*-----------------------------------------------------------------------------
+ * listpackEx functions
+ *----------------------------------------------------------------------------*/
+/*
+ * If 'hexpire' command is called on a listpack hash object for the first time,
+ * we convert it to OBJ_ENCODING_LISTPACK_EX encoding. Each field-value pair
+ * will have an additional TTL entry. So, listpack will have triplets in it:
+ * field-value-ttl. If TTL is not set for a field, we store 'zero' as the TTL
+ * value. 'zero' is encoded as two bytes in the listpack. Memory overhead of a
+ * non-existing TTL will be two bytes per field.
+ *
+ * Fields in the listpack will be ordered by TTL. Field with the smallest expiry
+ * time will be the first item. Fields without TTL will be at the end of the
+ * listpack. This way, it is easier/faster to find expired items.
+ */
+
+#define HASH_LP_NO_TTL 0
+
+static struct listpackEx *listpackExCreate(void) {
+    listpackEx *lpt = zcalloc(sizeof(*lpt));
+    lpt->meta.trash = 1;
+    lpt->lp = NULL;
+    lpt->key = NULL;
+    return lpt;
+}
+
+void *listpackExCreateFromListpack(void *lp) {
+    listpackEx *lpt = zcalloc(sizeof(*lpt));
+    lpt->lp = lp;
+    lpt->meta.trash = 1;
+    lpt->key = NULL;
+    return lpt;
+}
+
+void *listpackExGetListpack(const robj *o) {
+    listpackEx *lpt = o->ptr;
+    return lpt->lp;
+}
+
+void listpackExUpdateListpack(const robj *o, void *lp) {
+    listpackEx *lpt = o->ptr;
+    lpt->lp = lp;
+}
```

Comment:
since listpackEx can be accessed outside server.h, i think we can discard these methods, and listpackExCreateFromListpack() is only used in rdb.c, it's not a good deal to add a separate for it.

Reply:
OK, removed.
I created these in the first place because `listpackEx` (which was `listpackTTL` back then) was defined in `t_hash.c` and thus not accessible from `rdb.c`